

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>optwrf package &mdash; optwrf 1.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="optwrf.test" href="optwrf.test.html" />
    <link rel="prev" title="optwrf" href="modules.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> optwrf
          

          
          </a>

          
            
            
              <div class="version">
                1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Reference</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">optwrf</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">optwrf package</a><ul>
<li class="toctree-l3"><a class="reference internal" href="optwrf.test.html">optwrf.test</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-optwrf.linuxhelper">optwrf.linuxhelper</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-optwrf.optimize_wrf_physics">optwrf.optimize_wrf_physics</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-optwrf.runwrf">optwrf.runwrf</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-optwrf.simplega">optwrf.simplega</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-optwrf.wrfparams">optwrf.wrfparams</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">optwrf</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="modules.html">optwrf</a> &raquo;</li>
        
      <li>optwrf package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/optwrf.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="optwrf-package">
<h1>optwrf package<a class="headerlink" href="#optwrf-package" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="optwrf.test.html">optwrf.test</a><ul>
<li class="toctree-l2"><a class="reference internal" href="optwrf.test.html#module-optwrf.test.test_optimize_wrf_physics">optwrf.test.test_optimize_wrf_physics</a></li>
<li class="toctree-l2"><a class="reference internal" href="optwrf.test.html#module-optwrf.test.test_runwrf_postwrf">optwrf.test.test_runwrf_postwrf</a></li>
<li class="toctree-l2"><a class="reference internal" href="optwrf.test.html#module-optwrf.test.test_runwrf_setup">optwrf.test.test_runwrf_setup</a></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="module-optwrf.linuxhelper">
<span id="optwrf-linuxhelper"></span><h2>optwrf.linuxhelper<a class="headerlink" href="#module-optwrf.linuxhelper" title="Permalink to this headline">¶</a></h2>
<p>Functions that simplify running in a linux environment.
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
Known Issues/Wishlist:</p>
<dl class="function">
<dt id="optwrf.linuxhelper.remove_dir">
<code class="sig-prename descclassname">optwrf.linuxhelper.</code><code class="sig-name descname">remove_dir</code><span class="sig-paren">(</span><em class="sig-param">directory</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.linuxhelper.remove_dir" title="Permalink to this definition">¶</a></dt>
<dd><p>This function utilized an exception clause to delete a directory.</p>
<dl class="simple">
<dt>directory<span class="classifier">string</span></dt><dd><p>complete path to the directory to be removed.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-optwrf.optimize_wrf_physics">
<span id="optwrf-optimize-wrf-physics"></span><h2>optwrf.optimize_wrf_physics<a class="headerlink" href="#module-optwrf.optimize_wrf_physics" title="Permalink to this headline">¶</a></h2>
<p>A framework to run the simple genetic algorithm for optimizing the WRF model physics.
All simulation parameters and fitness values are saved in an SQL database.
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
Known Issues/Wishlist:
- This whole things needs to change…</p>
<blockquote>
<div><ul class="simple">
<li><p>I want the thread pool executor to operate within a Class or a function.</p></li>
</ul>
</div></blockquote>
<dl class="function">
<dt id="optwrf.optimize_wrf_physics.close_conn_to_db">
<code class="sig-prename descclassname">optwrf.optimize_wrf_physics.</code><code class="sig-name descname">close_conn_to_db</code><span class="sig-paren">(</span><em class="sig-param">db_conn</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.optimize_wrf_physics.close_conn_to_db" title="Permalink to this definition">¶</a></dt>
<dd><p>Closes the connection to a SQL database.</p>
<dl class="simple">
<dt>db_conn<span class="classifier">SQL database name</span></dt><dd><p>Can be ‘:memory:’ if you only want the database to be held in memory</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.optimize_wrf_physics.conn_to_db">
<code class="sig-prename descclassname">optwrf.optimize_wrf_physics.</code><code class="sig-name descname">conn_to_db</code><span class="sig-paren">(</span><em class="sig-param">db_name='optwrf.db'</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.optimize_wrf_physics.conn_to_db" title="Permalink to this definition">¶</a></dt>
<dd><p>Opens the connection to a SQL database.</p>
<dl class="simple">
<dt>db_name<span class="classifier">SQL database name</span></dt><dd><p>Can be ‘:memory:’ if you only want the database to be held in memory</p>
</dd>
</dl>
<p>db_conn : database connection object</p>
</dd></dl>

<dl class="function">
<dt id="optwrf.optimize_wrf_physics.get_fitness">
<code class="sig-prename descclassname">optwrf.optimize_wrf_physics.</code><code class="sig-name descname">get_fitness</code><span class="sig-paren">(</span><em class="sig-param">param_ids</em>, <em class="sig-param">start_date='Jan 15 2011'</em>, <em class="sig-param">end_date='Jan 16 2011'</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.optimize_wrf_physics.get_fitness" title="Permalink to this definition">¶</a></dt>
<dd><p>This function produces a random fitness value between 0 - 100</p>
<dl class="simple">
<dt>param_ids<span class="classifier">list of integers</span></dt><dd><p>Numeric values corresponding to each WRF physics parameterization.</p>
</dd>
</dl>
<dl class="simple">
<dt>fitness<span class="classifier">integer</span></dt><dd><p>A randomly generated integer between 1 and 100.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.optimize_wrf_physics.get_individual_by_genes">
<code class="sig-prename descclassname">optwrf.optimize_wrf_physics.</code><code class="sig-name descname">get_individual_by_genes</code><span class="sig-paren">(</span><em class="sig-param">individual</em>, <em class="sig-param">db_conn</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.optimize_wrf_physics.get_individual_by_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Looks for an indivual set of genes in an SQLite database.</p>
<dl class="simple">
<dt>individual<span class="classifier">Chromosome Class instance</span></dt><dd><p>lorem ipsum</p>
</dd>
</dl>
<dl class="simple">
<dt>past_sim<span class="classifier">simplega.Chromosome</span></dt><dd><p>Simulation that was run previously and stored in the SQLite database.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.optimize_wrf_physics.get_wrf_fitness">
<code class="sig-prename descclassname">optwrf.optimize_wrf_physics.</code><code class="sig-name descname">get_wrf_fitness</code><span class="sig-paren">(</span><em class="sig-param">param_ids</em>, <em class="sig-param">start_date='Jan 15 2011'</em>, <em class="sig-param">end_date='Jan 16 2011'</em>, <em class="sig-param">bc_data='ERA'</em>, <em class="sig-param">n_domains=1</em>, <em class="sig-param">correction_factor=0.00033806364898425066</em>, <em class="sig-param">setup_yaml='dirpath.yml'</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.optimize_wrf_physics.get_wrf_fitness" title="Permalink to this definition">¶</a></dt>
<dd><p>Using the input physics parameters, date, boundary condition, and domain data,
this function runs the WRF model and computes the error between WRF and ERA5.</p>
<dl class="simple">
<dt>param_ids<span class="classifier">list of integers</span></dt><dd><p>Numeric values corresponding to each WRF physics parameterization.</p>
</dd>
<dt>start_date<span class="classifier">string</span></dt><dd><p>lorem ipsum</p>
</dd>
<dt>end_date<span class="classifier">string</span></dt><dd><p>lorem ipsum</p>
</dd>
<dt>bc_data<span class="classifier">string</span></dt><dd><p>lorem ipsum</p>
</dd>
<dt>n_domains<span class="classifier">integer</span></dt><dd><p>lorem ipsum</p>
</dd>
<dt>setup_yaml<span class="classifier">string</span></dt><dd><p>lorem ipsum</p>
</dd>
</dl>
<dl class="simple">
<dt>fitness<span class="classifier">float</span></dt><dd><p>Fitness value denoting how well the WRF model run performed.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.optimize_wrf_physics.insert_sim">
<code class="sig-prename descclassname">optwrf.optimize_wrf_physics.</code><code class="sig-name descname">insert_sim</code><span class="sig-paren">(</span><em class="sig-param">individual</em>, <em class="sig-param">db_conn</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.optimize_wrf_physics.insert_sim" title="Permalink to this definition">¶</a></dt>
<dd><p>Inserts a simulation into the SQL database held in memory or written to a .db file.</p>
<dl class="simple">
<dt>individual<span class="classifier">simplega.Chromosome instance</span></dt><dd><p>The simulation that you would like to add to the simulation database.</p>
</dd>
<dt>db_conn :</dt><dd><p>lorem ipsum</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.optimize_wrf_physics.print_database">
<code class="sig-prename descclassname">optwrf.optimize_wrf_physics.</code><code class="sig-name descname">print_database</code><span class="sig-paren">(</span><em class="sig-param">db_conn</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.optimize_wrf_physics.print_database" title="Permalink to this definition">¶</a></dt>
<dd><p>Prints the entire SQLite simulation database.</p>
<dl class="simple">
<dt>db_conn :</dt><dd><p>lorem ipsum</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.optimize_wrf_physics.run_simplega">
<code class="sig-prename descclassname">optwrf.optimize_wrf_physics.</code><code class="sig-name descname">run_simplega</code><span class="sig-paren">(</span><em class="sig-param">pop_size</em>, <em class="sig-param">n_generations</em>, <em class="sig-param">testing=False</em>, <em class="sig-param">intial_pop=None</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.optimize_wrf_physics.run_simplega" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs the simple genetic algorithm specified in simplega either
to optimize the WRF model physics or with a test fitness function
that randomly selects a fitness value for each individual Chromosome.</p>
<p>I’m not sure if it would be better to make this a Class…</p>
<dl class="simple">
<dt>pop_size<span class="classifier">int</span></dt><dd><p>Desired population size for the genetic algorithm</p>
</dd>
<dt>n_generations<span class="classifier">int</span></dt><dd><p>Total number of generations after which the genetic algorithm will time out.</p>
</dd>
<dt>testing<span class="classifier">boolean</span></dt><dd><p>Flag that uses a random number generator as the fitness function when True.</p>
</dd>
</dl>
<dl class="simple">
<dt>WRFga_winner<span class="classifier">simplega.Chromosome instance</span></dt><dd><p>The simulation preforming the best in the genetic algorithm.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.optimize_wrf_physics.seed_initial_population">
<code class="sig-prename descclassname">optwrf.optimize_wrf_physics.</code><code class="sig-name descname">seed_initial_population</code><span class="sig-paren">(</span><em class="sig-param">input_csv</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.optimize_wrf_physics.seed_initial_population" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads the input csv file, which contains the dates and/or parameter combinations
that you would like to see in the initial population. The CSV file should be
formatted as follows:</p>
<dl class="simple">
<dt>Line 1: start_date, mp_physics, ra_lw_physics, ra_sw_physics, sf_surface_physics,</dt><dd><p>bl_pbl_physics, cu_physics, sf_sfclay_physics</p>
</dd>
</dl>
<p>Line 2: Feb 28 2011,  1, 24, 99,  4, 11,  2,  1
Line 3: Aug 31 2011, 28,  4,  5,  2, 10, 93, 10
Line 4: Dec 10 2011,  7, 24,  3,  4,  9,  4,  1
Line 5: Jul 22 2011, 11, 24,  2,  1,  4, 10,  4
Line 6: May 23 2011,  1,  4,  5,  2, 10, 93, 10
Line 7: Mar 31 2011, 28, 24, 99,  4, 11,  2,  1
.
.
.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>input_csv</strong> – string
defines the file name from which dates/parameters will be read.</p>
</dd>
<dt class="field-even">Return intial population</dt>
<dd class="field-even"><p>list of Chromosome instances
made from the information specified in the input file.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.optimize_wrf_physics.sql_to_csv">
<code class="sig-prename descclassname">optwrf.optimize_wrf_physics.</code><code class="sig-name descname">sql_to_csv</code><span class="sig-paren">(</span><em class="sig-param">csv_file_path</em>, <em class="sig-param">db_conn</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.optimize_wrf_physics.sql_to_csv" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes the contents of a SQL database to a CSV file. The SQL database must
be in the directory that you are running this function from.</p>
<dl class="simple">
<dt>csv_file_path<span class="classifier">string</span></dt><dd><p>Exact path to where you would like to csv to be saved ending with the file name.
e.g., csv_file_path = ‘/home/jas983/data/test.csv’</p>
</dd>
</dl>
<p>db_conn : database connection object</p>
</dd></dl>

<dl class="function">
<dt id="optwrf.optimize_wrf_physics.update_sim">
<code class="sig-prename descclassname">optwrf.optimize_wrf_physics.</code><code class="sig-name descname">update_sim</code><span class="sig-paren">(</span><em class="sig-param">individual</em>, <em class="sig-param">db_conn</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.optimize_wrf_physics.update_sim" title="Permalink to this definition">¶</a></dt>
<dd><p>Inserts a simulation into the SQL database held in memory or written to a .db file.</p>
<dl class="simple">
<dt>individual<span class="classifier">simplega.Chromosome instance</span></dt><dd><p>The simulation that you would like to add to the simulation database.</p>
</dd>
<dt>db_conn :</dt><dd><p>lorem ipsum</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-optwrf.runwrf">
<span id="optwrf-runwrf"></span><h2>optwrf.runwrf<a class="headerlink" href="#module-optwrf.runwrf" title="Permalink to this headline">¶</a></h2>
<p>Class and supporting functions to run WRF within other scripts
and process WRF output data.</p>
<p>Known Issues/Wishlist:
- I’m unhappy with the output to screen from rda_download(). Perhaps edit that and
check_file_status() as well.
- I want to figure out a better way to set the command aliai.
- Better Documentation</p>
<dl class="class">
<dt id="optwrf.runwrf.WRFModel">
<em class="property">class </em><code class="sig-prename descclassname">optwrf.runwrf.</code><code class="sig-name descname">WRFModel</code><span class="sig-paren">(</span><em class="sig-param">param_ids</em>, <em class="sig-param">start_date</em>, <em class="sig-param">end_date</em>, <em class="sig-param">bc_data='ERA'</em>, <em class="sig-param">n_domains=1</em>, <em class="sig-param">setup_yaml='dirpath.yml'</em>, <em class="sig-param">verbose=True</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.runwrf.WRFModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>This class provides a framework for running the WRF model</p>
<dl class="method">
<dt id="optwrf.runwrf.WRFModel.get_bc_data">
<code class="sig-name descname">get_bc_data</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.runwrf.WRFModel.get_bc_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads boundary condition data from the RDA
if it does not already exist in the expected data directory (self.DIR_DATA).
Then, it creates a temporary data directory (self.DIR_DATA_TMP) and copies
data from the central data directory to the temporary data directory.</p>
<p>The justificaion for this approach is that the link_grib.csh script doesn’t
contain a good way of picking files out of many. Therefore, it is much
easier to link all the files in the temporary data directory than to link
indidivdual files out of the central data directory.</p>
<p>NOTE Currently, only ERA data (ds627.0) is supported!</p>
<dl class="field-list simple">
<dt class="field-odd">Return vtable_sfx</dt>
<dd class="field-odd"><p>string
WPS variable table suffix – tells subsequent methods which boundary condidtion data is being used
so that ungrib can successfully unpack data.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="optwrf.runwrf.WRFModel.prepare_namelists">
<code class="sig-name descname">prepare_namelists</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.runwrf.WRFModel.prepare_namelists" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes dates, the geographical data path, number of domains, runtime duration,
and physics options to the WPS and/or WRF namelist files.</p>
</dd></dl>

<dl class="method">
<dt id="optwrf.runwrf.WRFModel.process_era5_data">
<code class="sig-name descname">process_era5_data</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.runwrf.WRFModel.process_era5_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads ERA5 data from the Research Data Archive if it doesn’t already exist in ERA5_ROOT,
and processes the data – reduces the size of the ERA5 files using ncks and
calculates the wind power density at 100m and the GHI – and writes
these variables to a montly ERA5 NetCDF data file named ERA5_EastUS_WPD-GHI_YYYY-MM.nc.</p>
<p>Unfortunately, the ERA5 surface solar radiation downward (ssrd) data is provided in incredibly
inconvenient temporal windows, so to build a montly file, we must download three ssrd files
and two wind files (U and V wind components). For example, the following filelist would be
built for January 2011:</p>
<dl>
<dt>filelist = [‘e5.oper.an.sfc/201101/</dt><dd><blockquote>
<div><p>e5.oper.an.sfc.228_246_100u.ll025sc.2011010100_2011013123.nc’,</p>
</div></blockquote>
<dl class="simple">
<dt>‘e5.oper.an.sfc/201101/</dt><dd><p>e5.oper.an.sfc.228_247_100v.ll025sc.2011010100_2011013123.nc’,</p>
</dd>
<dt>‘e5.oper.fc.sfc.accumu/201012/</dt><dd><p>e5.oper.fc.sfc.accumu.128_169_ssrd.ll025sc.2010121606_2011010106.nc’,</p>
</dd>
<dt>‘e5.oper.fc.sfc.accumu/201101/</dt><dd><p>e5.oper.fc.sfc.accumu.128_169_ssrd.ll025sc.2011010106_2011011606.nc’,</p>
</dd>
<dt>‘e5.oper.fc.sfc.accumu/201101/</dt><dd><p>e5.oper.fc.sfc.accumu.128_169_ssrd.ll025sc.2011011606_2011020106.nc’]</p>
</dd>
</dl>
</dd>
</dl>
<p>which means there are three separate directories on RDA that files must be downloaded from.</p>
<p><a href="#id1"><span class="problematic" id="id2">**</span></a><a href="#id3"><span class="problematic" id="id4">*</span></a>NOTE that this method currently only supports simulations within a single month
i.e., YYYY-MM-01 00:00:00 - YYYY-MM-&lt;end day&gt; 23:00:00. Otherwise, wrf_era5_diff()
will fail. Times are in UTC.</p>
<p><a href="#id5"><span class="problematic" id="id6">**</span></a><a href="#id7"><span class="problematic" id="id8">*</span></a>NOTE that this method does not currently support running on Cheyenne;
it’s about halfway there…</p>
</dd></dl>

<dl class="method">
<dt id="optwrf.runwrf.WRFModel.process_wrfout_data">
<code class="sig-name descname">process_wrfout_data</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.runwrf.WRFModel.process_wrfout_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Processes the wrfout file – calculates GHI and wind power denity (WPD) and writes these variables
to wrfout_processed_d01.nc data file to be used by the regridding script (wrf2era_error.ncl) in
wrf_era5_diff().</p>
<p>NEEDS BETTER DOCUMENTATION!!!</p>
</dd></dl>

<dl class="method">
<dt id="optwrf.runwrf.WRFModel.run_real">
<code class="sig-name descname">run_real</code><span class="sig-paren">(</span><em class="sig-param">disable_timeout=False</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.runwrf.WRFModel.run_real" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs real.exe and checks to see that if it was successful.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>boolean (True/False)
If runwrf_finish_check for real returns ‘complete’ (‘failed’),
this function returns True (False).</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="optwrf.runwrf.WRFModel.run_wps">
<code class="sig-name descname">run_wps</code><span class="sig-paren">(</span><em class="sig-param">disable_timeout=False</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.runwrf.WRFModel.run_wps" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs the WRF preprocessing executables and confirms thier success.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>boolean (True/False)
If runwrf_finish_check for geogrid and metgrid
returns ‘complete’ (‘failed’), this function returns True (False).</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="optwrf.runwrf.WRFModel.run_wrf">
<code class="sig-name descname">run_wrf</code><span class="sig-paren">(</span><em class="sig-param">disable_timeout=False</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.runwrf.WRFModel.run_wrf" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs wrf.exe and checks to see if it was successful.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>boolean (True/False)
If runwrf_finish_check for wrf returns ‘complete’ (‘failed’),
this function returns True (False).</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="optwrf.runwrf.WRFModel.runwrf_finish_check">
<code class="sig-name descname">runwrf_finish_check</code><span class="sig-paren">(</span><em class="sig-param">program</em>, <em class="sig-param">nprocs=8</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.runwrf.WRFModel.runwrf_finish_check" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if a specified WRF subprogram has finished running. Does this specifically by
checking the first and the last rsl.out.* file created by each processor running
either wrf.exe or real.exe via OpenMPI.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>program</strong> – string
WRF or WPS subprogram name whose status is to be checked.</p></li>
<li><p><strong>nprocs</strong> – integer
Number of processors that you are using to run real.exe and wrf.exe.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>‘running’ or ‘complete’ or ‘failed’ string
Run status of the program.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="optwrf.runwrf.WRFModel.wrf_era5_diff">
<code class="sig-name descname">wrf_era5_diff</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.runwrf.WRFModel.wrf_era5_diff" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the difference between the wrf simulation and ERA5
by calling the wrf2era_error.ncl script from the command line.
NCL (short for NCAR Command Language) will be deprecated at
some point in the future, and will be replaced by wrf-python,
but the regridding functionality has not been ported to this
package currently. Therefore, this function is a somewhat clumsy
workaround.</p>
<p>wrf2era_error.ncl reads the previously-processed ERA and wrfout files,
regrids the wrfout global horizontal irradiance (GHI) and wind
power density (WPD) to match that of the ERA data, and then computes
the difference across the entire d01 WRF domain for every time period.</p>
<p>This method waits while the regridding and error calculation is being
completed by NCL, which can take some seconds to complete.</p>
<p><a href="#id9"><span class="problematic" id="id10">**</span></a><a href="#id11"><span class="problematic" id="id12">*</span></a>NOTE: I don’t need to copy the wrf2era_error.ncl script into each
WRFOUT directory, but that’s how I currently get the error file for
each simulation to be written into the correct dirctory.</p>
<dl class="field-list simple">
<dt class="field-odd">Return total_error</dt>
<dd class="field-odd"><p>list
Sum of the mean absolute error accumulated for each grid cell
during each time period in the WRF simulation.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="optwrf.runwrf.WRFModel.wrfdir_setup">
<code class="sig-name descname">wrfdir_setup</code><span class="sig-paren">(</span><em class="sig-param">vtable_sfx</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.runwrf.WRFModel.wrfdir_setup" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets up the WRF run directory by copying scripts, data files, and executables.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>vtable_sfx</strong> – string
variable table suffix specific to the boundary condition data;
required to run WPS.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="optwrf.runwrf.check_file_status">
<code class="sig-prename descclassname">optwrf.runwrf.</code><code class="sig-name descname">check_file_status</code><span class="sig-paren">(</span><em class="sig-param">filepath</em>, <em class="sig-param">filesize</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.runwrf.check_file_status" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks the file status during a download from the internet.</p>
<p>This is currently not implemented because I don’t
like the way it prints information to the command line.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>filepath</strong> – string
Path to remote data file</p>
</dd>
</dl>
<dl class="simple">
<dt>:param filesize<span class="classifier">int</span></dt><dd><p>Size of the file as found by req.headers[‘Content-length’]</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.runwrf.determine_computer">
<code class="sig-prename descclassname">optwrf.runwrf.</code><code class="sig-name descname">determine_computer</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.runwrf.determine_computer" title="Permalink to this definition">¶</a></dt>
<dd><p>Determines which computer you are currently working on.</p>
<dl class="field-list simple">
<dt class="field-odd">Return on_aws</dt>
<dd class="field-odd"><p>boolean
True if working on the mzhang AWS account where the group name is ‘ec2-user’</p>
</dd>
<dt class="field-even">Return on_cheyenne</dt>
<dd class="field-even"><p>boolean
True if working on the NCAR Cheyenne super computer where the group name is ‘ncar’</p>
</dd>
<dt class="field-odd">Return on_magma</dt>
<dd class="field-odd"><p>boolean
True if working on Jeff Sward’s account on the Cornell Magma cluster where the group name is ‘pug-jas983’</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.runwrf.format_date">
<code class="sig-prename descclassname">optwrf.runwrf.</code><code class="sig-name descname">format_date</code><span class="sig-paren">(</span><em class="sig-param">in_date</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.runwrf.format_date" title="Permalink to this definition">¶</a></dt>
<dd><p>Formats an input date so that it can be correctly written to the namelist.</p>
<dl class="simple">
<dt>:param in_date<span class="classifier">string</span></dt><dd><p>string specifying the date</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>datetime64 array specifying the date</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.runwrf.print_last_3lines">
<code class="sig-prename descclassname">optwrf.runwrf.</code><code class="sig-name descname">print_last_3lines</code><span class="sig-paren">(</span><em class="sig-param">file_name</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.runwrf.print_last_3lines" title="Permalink to this definition">¶</a></dt>
<dd><p>Prints the last three lines of a file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>file_name</strong> – string
Complete path of the file that you would like print.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.runwrf.rda_download">
<code class="sig-prename descclassname">optwrf.runwrf.</code><code class="sig-name descname">rda_download</code><span class="sig-paren">(</span><em class="sig-param">filelist</em>, <em class="sig-param">dspath</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.runwrf.rda_download" title="Permalink to this definition">¶</a></dt>
<dd><p>Logs into the NCAR research data archive (RDA)
and downloads specified files.</p>
<p>NOTE: My username/password are currently hard-coded into this function.
I SHOULD CHAGE THIS!</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>filelist</strong> – list of strings
List of all the files that you would like downloaded from the RDA.</p>
</dd>
</dl>
<dl class="simple">
<dt>:param dspath<span class="classifier">string</span></dt><dd><p>Full path to file on the RDA. You can obtain this from</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a boolean (True/False) success flag.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.runwrf.read_2nd2_last_line">
<code class="sig-prename descclassname">optwrf.runwrf.</code><code class="sig-name descname">read_2nd2_last_line</code><span class="sig-paren">(</span><em class="sig-param">file_name</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.runwrf.read_2nd2_last_line" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads the second to last line of a file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>file_name</strong> – string
Complete path of the file that you would like read.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>second2_last_line: string
Second to last line of the input file.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.runwrf.read_last_3lines">
<code class="sig-prename descclassname">optwrf.runwrf.</code><code class="sig-name descname">read_last_3lines</code><span class="sig-paren">(</span><em class="sig-param">file_name</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.runwrf.read_last_3lines" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads the last three lines of a file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>file_name</strong> – string
Complete path of the file that you would like print.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>last_3lines: string
Last three lines of the input file.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.runwrf.read_last_line">
<code class="sig-prename descclassname">optwrf.runwrf.</code><code class="sig-name descname">read_last_line</code><span class="sig-paren">(</span><em class="sig-param">file_name</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.runwrf.read_last_line" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads the last line of a file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>file_name</strong> – string
Complete path of the file that you would like read.</p>
</dd>
<dt class="field-even">Return last_line</dt>
<dd class="field-even"><p>string
Last line of the input file.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-optwrf.simplega">
<span id="optwrf-simplega"></span><h2>optwrf.simplega<a class="headerlink" href="#module-optwrf.simplega" title="Permalink to this headline">¶</a></h2>
<p>A simple genetic algorithm for optimizing the WRF model physics.</p>
<p>Known Issues/Wishlist:</p>
<dl class="class">
<dt id="optwrf.simplega.Chromosome">
<em class="property">class </em><code class="sig-prename descclassname">optwrf.simplega.</code><code class="sig-name descname">Chromosome</code><span class="sig-paren">(</span><em class="sig-param">genes</em>, <em class="sig-param">start_date</em>, <em class="sig-param">end_date</em>, <em class="sig-param">fitness=None</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.simplega.Chromosome" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>This class provides information about each unique individual to be placed within
populations used in the genetic algorithm (e.g., parent population, mating population,
or offspring population). Four pieces of information are associated with each individual:</p>
<blockquote>
<div><p>1. A set of genes corresponding to six major physics parameters that be changed within
the WRF model (microphysics, longwave radiation, shortwave radiation, land surface,
planetary boundary layer, and cumulus).
2. A start date defining when the WRF forecast will begin running.
3. An end date defining when the WRF forecast will finish running.
4. A fitness value that will be used to determine how well each instance preforms.</p>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="optwrf.simplega.crossover">
<code class="sig-prename descclassname">optwrf.simplega.</code><code class="sig-name descname">crossover</code><span class="sig-paren">(</span><em class="sig-param">mating_population</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.simplega.crossover" title="Permalink to this definition">¶</a></dt>
<dd><p>The crossover operator takes in the genes of two parent individuals
and randomly exchanges one gene between the two producing two offspring.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>mating_population</strong> – list of Chromosome instances
population of individuals that have been deemed fit enough by the selection operator.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>offspring_population: list of Chromosome instances
population of individuals created from the genes of mating population.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.simplega.display">
<code class="sig-prename descclassname">optwrf.simplega.</code><code class="sig-name descname">display</code><span class="sig-paren">(</span><em class="sig-param">individual</em>, <em class="sig-param">start_time</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.simplega.display" title="Permalink to this definition">¶</a></dt>
<dd><p>Aids in display of an individual Chromosome instanceby printing useful information
to the screen.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>individual</strong> – Chromosome instance</p></li>
<li><p><strong>start_time</strong> – datetime object
defining when the simiulation began running.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.simplega.display_pop">
<code class="sig-prename descclassname">optwrf.simplega.</code><code class="sig-name descname">display_pop</code><span class="sig-paren">(</span><em class="sig-param">population</em>, <em class="sig-param">fn_display</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.simplega.display_pop" title="Permalink to this definition">¶</a></dt>
<dd><p>Aids in display of an entire population by printing useful information
to the screen.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>population</strong> – list of Chromosome instances</p></li>
<li><p><strong>fn_display</strong> – function
defining how to display eah individual in the population.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.simplega.find_elites">
<code class="sig-prename descclassname">optwrf.simplega.</code><code class="sig-name descname">find_elites</code><span class="sig-paren">(</span><em class="sig-param">population</em>, <em class="sig-param">n_elites</em>, <em class="sig-param">fn_display_pop</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.simplega.find_elites" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds elites within the population by sorting the members of the population from
best (numerically lowest) to worst (numerically highest) fitness value,
and returns a list (with the nuber of individuals specified by the user)
containing the individuals with the best fitness.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>population</strong> – list of Chromosome instances</p></li>
<li><p><strong>n_elites</strong> – integer
specifying the number of individuals you want to keep from the population.</p></li>
<li><p><strong>fn_display_pop</strong> – function
for printing information about the population to the screen.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>elites: list of Chromosome instances
specifying the elite individuals extracted from the population.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.simplega.generate_genes">
<code class="sig-prename descclassname">optwrf.simplega.</code><code class="sig-name descname">generate_genes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.simplega.generate_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates genes for a new individual to be placed within a population.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
<dl class="simple">
<dt>new individual: list</dt><dd><p>list of WRF model physics parameters</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.simplega.generate_population">
<code class="sig-prename descclassname">optwrf.simplega.</code><code class="sig-name descname">generate_population</code><span class="sig-paren">(</span><em class="sig-param">pop_size</em>, <em class="sig-param">initial_population=None</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.simplega.generate_population" title="Permalink to this definition">¶</a></dt>
<dd><p>This method provides a way to randomly generate an initial population for the GA.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pop_size</strong> – integer
specifies the number of individuals in the population to be generated.</p></li>
<li><p><strong>initial_population</strong> – list of Chromosome instances
allows users to seed the population with individuals of desired
parameter combinations and start/end dates.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>population: list of Chromosome instances</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.simplega.generate_random_dates">
<code class="sig-prename descclassname">optwrf.simplega.</code><code class="sig-name descname">generate_random_dates</code><span class="sig-paren">(</span><em class="sig-param">year=2011</em>, <em class="sig-param">n_days=1</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.simplega.generate_random_dates" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a random start date and end date for running the WRF model.
Start and end dates are offset by the parameter n_days, which is set to 1 as default</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>year</strong> – integer or None
specifying the year within which you want random days to be selected.
If this is specified as None, random dates between Jan 1, 2000 and today can be selected.</p></li>
<li><p><strong>n_days</strong> – integer
specifying the nuber of days between the start and the end dates of the simulation.
The default is set to 1 day.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<dl class="simple">
<dt>random_start_date: string</dt><dd><p>specifys the simulation start date</p>
</dd>
<dt>random_end_date: string</dt><dd><p>specifys the simulation end date</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.simplega.get_best">
<code class="sig-prename descclassname">optwrf.simplega.</code><code class="sig-name descname">get_best</code><span class="sig-paren">(</span><em class="sig-param">population</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.simplega.get_best" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds the location of the individual within the population with the best (lowest) fitness.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>population</strong> – list of Chromosome instances</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>best_individual: single Chromosome instance
individual within a population with the lowest fitness.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.simplega.mutate">
<code class="sig-prename descclassname">optwrf.simplega.</code><code class="sig-name descname">mutate</code><span class="sig-paren">(</span><em class="sig-param">offspring_population</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.simplega.mutate" title="Permalink to this definition">¶</a></dt>
<dd><p>The mutate operator takes in the genes of one offspring
and randomly replaces one of the genes from the set of choices.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>offspring_population</strong> – list of Chromosome instances
population of individuals created from the genes of mating population.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>offspring_population: list of Chromosome instances
mutated offspring population.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.simplega.selection">
<code class="sig-prename descclassname">optwrf.simplega.</code><code class="sig-name descname">selection</code><span class="sig-paren">(</span><em class="sig-param">population</em>, <em class="sig-param">pop_size</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.simplega.selection" title="Permalink to this definition">¶</a></dt>
<dd><p>The selection operator carries out a tournament selection,
by selecting the fittest individual among a group of randomly selected individuals,
to create the mating population.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>population</strong> – list of Chromosome instances
specifies the population that the selection operator will select from.</p></li>
<li><p><strong>pop_size</strong> – integer
number of individuals within the population</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>mating population: list of Chromosome instances
population of individuals that have been deemed fit enough by the selection operator
and will be sent to the crossover operator.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-optwrf.wrfparams">
<span id="optwrf-wrfparams"></span><h2>optwrf.wrfparams<a class="headerlink" href="#module-optwrf.wrfparams" title="Permalink to this headline">¶</a></h2>
<p>Generates sets of WRF physics parameters, finds corresponding numeric identifies,
and writes the numeric paramter options to a CSV.
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
Known Issues/Wishlist:
- It would be better if I could read in the parameters from the in_yaml
file only once, but I’m not sure how variables across different
functions in the same module work.</p>
<dl class="function">
<dt id="optwrf.wrfparams.apply_dependencies">
<code class="sig-prename descclassname">optwrf.wrfparams.</code><code class="sig-name descname">apply_dependencies</code><span class="sig-paren">(</span><em class="sig-param">param_ids</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.wrfparams.apply_dependencies" title="Permalink to this definition">¶</a></dt>
<dd><p>Applies depependencies among parameters discovered during runtime</p>
<dl class="simple">
<dt>param_ids: list</dt><dd><p>List of WRF physics parameters</p>
</dd>
</dl>
<dl class="simple">
<dt>param_ids: list</dt><dd><p>List of WRF physics parameters after known dependencies have been applied</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="optwrf.wrfparams.combine">
<code class="sig-prename descclassname">optwrf.wrfparams.</code><code class="sig-name descname">combine</code><span class="sig-paren">(</span><em class="sig-param">lst1</em>, <em class="sig-param">lst2</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.wrfparams.combine" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="optwrf.wrfparams.filldefault">
<code class="sig-prename descclassname">optwrf.wrfparams.</code><code class="sig-name descname">filldefault</code><span class="sig-paren">(</span><em class="sig-param">in_yaml</em>, <em class="sig-param">in_param_ids</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.wrfparams.filldefault" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="optwrf.wrfparams.flexible_generate">
<code class="sig-prename descclassname">optwrf.wrfparams.</code><code class="sig-name descname">flexible_generate</code><span class="sig-paren">(</span><em class="sig-param">generate_params=True</em>, <em class="sig-param">mp=None</em>, <em class="sig-param">lw=None</em>, <em class="sig-param">sw=None</em>, <em class="sig-param">lsm=None</em>, <em class="sig-param">pbl=None</em>, <em class="sig-param">cu=None</em>, <em class="sig-param">in_yaml='params.yml'</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.wrfparams.flexible_generate" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a parameter combination of the 6 core parameters if the user has specified this option.
Otherwise, use specified input parameters and use defaults for the remaining paramters.</p>
</dd></dl>

<dl class="function">
<dt id="optwrf.wrfparams.generate">
<code class="sig-prename descclassname">optwrf.wrfparams.</code><code class="sig-name descname">generate</code><span class="sig-paren">(</span><em class="sig-param">in_yaml='params.yml'</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.wrfparams.generate" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a random set of parameters.</p>
</dd></dl>

<dl class="function">
<dt id="optwrf.wrfparams.ids2str">
<code class="sig-prename descclassname">optwrf.wrfparams.</code><code class="sig-name descname">ids2str</code><span class="sig-paren">(</span><em class="sig-param">param_ids</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.wrfparams.ids2str" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="optwrf.wrfparams.name2num">
<code class="sig-prename descclassname">optwrf.wrfparams.</code><code class="sig-name descname">name2num</code><span class="sig-paren">(</span><em class="sig-param">in_yaml='params.yml'</em>, <em class="sig-param">use_defaults=True</em>, <em class="sig-param">mp_in='morrison2mom'</em>, <em class="sig-param">lw_in='rrtm'</em>, <em class="sig-param">sw_in='dudia'</em>, <em class="sig-param">lsm_in='noah'</em>, <em class="sig-param">pbl_in='myj'</em>, <em class="sig-param">clo_in='grell-freitas'</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.wrfparams.name2num" title="Permalink to this definition">¶</a></dt>
<dd><p>Maps the name of the parameterization,
which is defined generally by the name of the individual or
institution that designed the scheme, to the namelist.input value.
The default value of these parameterization schemes is set by those that
were originally used by ICF in the study over NYC.</p>
</dd></dl>

<dl class="function">
<dt id="optwrf.wrfparams.pbl2sfclay">
<code class="sig-prename descclassname">optwrf.wrfparams.</code><code class="sig-name descname">pbl2sfclay</code><span class="sig-paren">(</span><em class="sig-param">id_pbl</em>, <em class="sig-param">rnd=False</em><span class="sig-paren">)</span><a class="headerlink" href="#optwrf.wrfparams.pbl2sfclay" title="Permalink to this definition">¶</a></dt>
<dd><p>Assigns the surface layer scheme based on the
specified or randomly selected PBL scheme. If multiple surface layer
schemes are available, one may be selected at random by setting rnd = True.
Otherwise, sf_sfclay defaults to option 1 (the revised MM5 scheme).</p>
</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="optwrf.test.html" class="btn btn-neutral float-right" title="optwrf.test" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="modules.html" class="btn btn-neutral float-left" title="optwrf" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Jeffrey A Sward

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>